router-tests/websocket_test.go

```diff
@@ -2159,3 +2166,236 @@ func expectConnectAndReadCurrentTime(t *testing.T, xEnv *testenv.Environment) {
+							return
+						}
+
+					case "subscribe":
+						// Handle countEmp subscription
+						if payload, ok := msg["payload"].(map[string]interface{}); ok {
+							if query, ok := payload["query"].(string); ok && strings.Contains(query, "countEmp") {
+								go handleCountEmpSubscription(t, wsWriteCh, msg["id"], 200*time.Millisecond, totalUpdates)
+							}
+						}
+
+					case "complete":
+						// Client completed subscription
+						return
+					}
+				}
+			})
+		}
+
+		// Configure and run the test
+		testenv.Run(t, &testenv.Config{
+			Subgraphs: testenv.SubgraphsConfig{
+				GlobalMiddleware: wsMiddleware,
+			},
+			ModifyEngineExecutionConfiguration: func(config *config.EngineExecutionConfiguration) {
+				// Use short ping interval to test quickly
+				config.WebSocketClientPingInterval = 100 * time.Millisecond
+			},
+		}, func(t *testing.T, xEnv *testenv.Environment) {
+			// Setup client connection
+			conn := xEnv.InitGraphQLWebSocketConnection(nil, nil, nil)
+
+			// Start the subscription
+			err := testenv.DeflakeWSWriteJSON(t, conn, testenv.WebSocketMessage{
+				ID:      "1",
+				Type:    "subscribe",
+				Payload: []byte(`{"query":"subscription { countEmp(max: 5, intervalMilliseconds: 200) }"}`),
+			})
+			require.NoError(t, err)
+
+			// Process subscription updates
+			var receivedUpdates int
+			for receivedUpdates < totalUpdates {
+				var res testenv.WebSocketMessage
+				err = testenv.DeflakeWSReadJSON(t, conn, &res)
+				require.NoError(t, err)
+
+				if res.Type == "next" {
+					receivedUpdates++
+					var payload struct {
+						Data struct {
+							CountEmp int `json:"countEmp"`
+						} `json:"data"`
+					}
+					err := json.Unmarshal(res.Payload, &payload)
+					require.NoError(t, err)
+					require.Equal(t, receivedUpdates, payload.Data.CountEmp)
+				}
+			}
+
+			// Get the complete message
+			var complete testenv.WebSocketMessage
+			err = testenv.DeflakeWSReadJSON(t, conn, &complete)
+			require.NoError(t, err)
+			require.Equal(t, "complete", complete.Type)
+
+			// Check that we received at least one ping
+			timeoutCh := time.After(3 * time.Second)
+			pingCount := 0
+
+			for {
+				select {
+				case <-pingReceived:
+					pingCount++
+
+					if pingCount >= totalUpdates {
+						t.Logf("Received %d pings from router", pingCount)
+						return
+					}
+				case <-timeoutCh:
+					require.GreaterOrEqual(t, pingCount, 1, "Expected at least one ping from router")
+					return
+				}
+			}
+		})
+
+	})
+}
+
+// Helper function to handle countEmp subscription
+func handleCountEmpSubscription(t *testing.T, wsWriteCh chan<- wsMessage, id interface{}, updateInterval time.Duration, totalUpdates int) {
+	// Send updates with the specified interval
+	for i := 1; i <= totalUpdates; i++ {
+		response := map[string]interface{}{
+			"type": "next",
+			"id":   id,
+			"payload": map[string]interface{}{
+				"data": map[string]interface{}{
+					"countEmp": i,
+				},
+			},
+		}
```

Comment:
This is the same type as the `msg` in the other part of the test right? I think this would be a good candidate for a struct.